vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,524 @@
+    pub sequence_number: u64,
+    pub file_write_guid: u128,
+    pub data_write_guid: u128,
+
+    pub log_guid: u128,
+    pub log_version: u16,
+    pub version: u16,
+    pub log_length: u32,
+    pub log_offset: u64,
+}
+
+impl Header {
+    /// Reads the Header structure from a reference VHDx file
+    pub fn new(f: &mut File, start: u64) -> Result<Header> {
+        // Read the whole header in to a buffer. We will need if for calculating checksum
+        let mut buffer = [0; HEADER_SIZE as usize];
+        f.seek(SeekFrom::Start(start)).map_err(Error::ReadHeader)?;
+        f.read_exact(&mut buffer).map_err(Error::ReadHeader)?;
+
+        let signature = LittleEndian::read_u32(&buffer[0..4]);
+        if signature != HEADER_SIGN {
+            return Err(Error::InvalidHeaderSign);
+        }
+
+        let checksum = LittleEndian::read_u32(&buffer[4..8]);
+        let new_checksum = calculate_checksum(&mut buffer, size_of::<u32>())?;
+        if checksum != new_checksum {
+            return Err(Error::InvalidChecksum(String::from("Header")));
+        }
+
+        Ok(Header {
+            signature,
+            checksum,
+            sequence_number: LittleEndian::read_u64(&buffer[8..16]),
+            file_write_guid: LittleEndian::read_u128(&buffer[16..32]),
+            data_write_guid: LittleEndian::read_u128(&buffer[32..48]),
+            log_guid: LittleEndian::read_u128(&buffer[48..64]),
+            log_version: LittleEndian::read_u16(&buffer[64..66]),
+            version: LittleEndian::read_u16(&buffer[66..68]),
+            log_length: LittleEndian::read_u32(&buffer[68..72]),
+            log_offset: LittleEndian::read_u64(&buffer[72..80]),
+        })
+    }
+
+    /// Converts the header structure in to a buffer
+    fn get_header_as_buffer(&self, buffer: &mut [u8; HEADER_SIZE as usize]) {
+        LittleEndian::write_u32(&mut buffer[0..4], self.signature);
+        LittleEndian::write_u32(&mut buffer[4..8], self.checksum);
+        LittleEndian::write_u64(&mut buffer[8..16], self.sequence_number);
+        LittleEndian::write_u128(&mut buffer[16..32], self.file_write_guid);
+        LittleEndian::write_u128(&mut buffer[32..48], self.data_write_guid);
+        LittleEndian::write_u128(&mut buffer[48..64], self.log_guid);
+        LittleEndian::write_u16(&mut buffer[64..66], self.log_version);
+        LittleEndian::write_u16(&mut buffer[66..68], self.version);
+        LittleEndian::write_u32(&mut buffer[68..72], self.log_length);
+        LittleEndian::write_u64(&mut buffer[72..80], self.log_offset);
+    }
+
+    /// Creates and returns new updated header from the provided current header
+    pub fn update_header(
+        f: &mut File,
+        current_header: Header,
+        change_data_guid: bool,
+        mut file_write_guid: u128,
+        start: u64,
+    ) -> Result<Header> {
+        let mut buffer = [0u8; HEADER_SIZE as usize];
+        let mut new_header: Header;
+        let data_write_guid;
```

Comment:
Not really idiomatic to Rust. You should declare these variables when they get filled with data.